doc/guides/porting/arch.rst

```diff
@@ -460,8 +460,97 @@ be derived from the linker scripts of other architectures. Some sections might
 be specific to the new architecture, for example the SCB section on ARM and the
 IDT section on x86.
 
-Hardware Stack Protection
-=========================
+Memory Management
+*****************
+
+If the target platform enables paging and requires drivers to memory-map
+their I/O regions, :option:`CONFIG_MMU` needs to be enabled and the
+:cpp:func:`arch_mem_map()` API implemented.
+
+Stack Objects
+*************
+
+The presence of memory protection hardware affects how stack objects are
+created. All architecture ports must specify the required alignment of the
+stack pointer, which is some combination of CPU and ABI requirements. This
+is defined in architecture headers with :c:macro:`ARCH_STACK_PTR_ALIGN` and
+is typically something small like 4, 8, or 16 bytes.
+
+Two types of thread stacks exist:
+
+- "kernel" stacks defined with :c:macro:`K_KERNEL_STACK_DEFINE()` and related
+  APIs, which can host kernel threads running in supervisor mode or
+  used as the stack for interrupt/exception handling. These have significantly
+  relaxed alignment requirements and use less reserved data.
+
+- "thread" stacks which typically use more memory, but are capable of hosting
+  thread running in user mode, as well as any use-cases for kernel stacks.
+
+If :c:option:`CONFIG_USERSPACE` is not enabled, "thread" and "kernel" stacks
+are equivalent.
+
+Additional macros may be defined in the architecture layer to specify
+the alignment of the base of stack objects, any reserved data inside the
+stack object not used for the thread's stack buffer, and how to round up
+stack sizes to support user mode threads. In the absence of definitions
+some defaults are assumed:
+
+- :c:macro:`ARCH_KERNEL_STACK_RESERVED`: default no reserved space
+- :c:macro:`ARCH_THREAD_STACK_RESERVED`: default no reserved space
+- :c:macro:`ARCH_KERNEL_STACK_OBJ_ALIGN`: default align to
+  :c:macro:`ARCH_STACK_PTR_ALIGN`
+- :c:macro:`ARCH_THREAD_STACK_OBJ_ALIGN`: default align to
+  :c:macro:`ARCH_STACK_PTR_ALIGN`
+- :c:macro:`ARCH_THREAD_STACK_SIZE_ALIGN(size)`: default round up to
+  :c:macro:`ARCH_STACK_PTR_ALIGN`
+
+All stack creation macros are defined in terms of these.
+
+Stack objects all have the following layout, with some regions potentially
+zero-sized depending on configuration. There are always two main parts:
+reserved memory at the beginning, and then the stack buffer itself. The
+bounds of some areas can only be determined at runtime in the context of
+its associated thread object. Other areas are entirely computable at build
+time.
+
+Some architectures may need to carve-out reserved memory at runtime from the
+stack buffer, instead of unconditionally reserving it at build time.
```

Comment:
In some use-cases, it is not "instead of" but "besides", right? Cause we do both. E.g. for FPU_SHARING in ARM

Reply:
I changed to "Some architectures may need to carve-out reserved memory at runtime from the
stack buffer, instead of unconditionally reserving it at build time, or to
supplement an existing reserved area (as is the case with the ARM FPU)."

> > Updated based on @ioannisg comments.
> > Also, based on an offline discussion with @dleach02, I've simplified ARM to always have K_THREAD_STACK_RESERVED be 0. It's simpler this way since ARM is unconditionally generating separate privilege stacks.
> 
> @andrewboie although, I am not against this simplification, I'd like to point out that now K_THREAD_STACK_RESERVED for ARMv8-M, will be 32 when CONFIG_USERSPACE=n and 0 when CONFIG_USERSPACE=y, while the HW does not impose any relevant constraints that justify this difference. This might be confusing for a developer that builds his code with / without USERSPACE enabled and sees different available stack for their _supervisor_ threads (assuming that they use K_THREAD_STACK_DEFINE).
> 
> We had always been treating the (CONFIG_USERSPACE && CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT) as _the_ special case.
> 
> If you do this simplification, it means you favor consistency across the different cortex-m variants.
> But maybe it is better to favor consistency across building with/without USERSPACE for each cortex-m variant.
> In both cases you need to sacrifice something.

I'd like to know what you would like me to do to move forward.

If we reserve data, then we have two guard regions, one in the privilege stack and one in the regular stack, with the regular stack guard wasted when the thread drops to user mode.

If we do not reserve data, then the situation is as described above.

IMO the right way to fix this is to convert ARM to not use privilege mode stack generation when it doesn't need to, which I believe we agreed was the next step.